File: Java/2020/EscapeRoom.java
// J5

import java.util.ArrayList;
import java.util.Scanner;

class Pair {
    int x, y;

    public Pair(int xval, int yval) {
        x = xval;
        y = yval;
    }

}

public class EscapeRoom {

    private static int down, across;
    private static boolean found = false;
    private static ArrayList<Pair> visited = new ArrayList<Pair>();
    private static int[][] matrix;

    public static ArrayList<Pair> factors(int value) {
        ArrayList<Pair> factors = new ArrayList<Pair>();

        for (int i = 1; i < value + 1; i++) {
            if(value % i == 0) {
                if (i <= down && (value / i) <= across) {
                    factors.add(new Pair(i, (int)(value/i)));
                }
            }
        }

        return factors;
    }    

    public static void check(int d, int a) {
        if(d == down && a == across) {
            found = true;
            return;
        }
        ArrayList<Pair> factors = factors(matrix[d-1][a-1]);

        if (!found) {
            for (Pair elem : factors) {
                if(!visited.contains(elem)) {
                    visited.add(elem);
                    check(elem.x, elem.y);
                }
            }
        }
        return;
    }

    public static void main(String[] args) {
        
        Scanner scanner = new Scanner(System.in);
        
        down = scanner.nextInt();
        across = scanner.nextInt();

        matrix = new int[down][across];

        for (int a = 0; a < down; a++) {
            for (int b = 0; b < across; b++) {
                matrix[a][b] = scanner.nextInt();
            }
        }

        check(1, 1);

        if (found) System.out.println("yes"); else System.out.println("no");
        
        scanner.close();
    }
}

File: Python/2020/Escape Room.py
# J5

found = False
visited = [] # stores all visited coordinates
matrix = [] # [3 10 8 14]
            # [1 11 12 12]
            # [6 2  3  9]


def factors(value):
    factors = []
    for i in range(1, value + 1):
        if value % i == 0:
            if i <= down and value / i <= across:
                factors.append((i, int(value / i)))
    return factors # [(1, 3), (3, 1)]


def check(d, a): # these parameters are real positions (3, 4)
    global found, matrix
    if d == down and a == across:
        found = True
        return
    facts = factors(matrix[d - 1][a - 1])
    if not found:
        for elem in facts:
            if elem not in visited:
                visited.append(elem)
                check(elem[0], elem[1])
    return


if __name__ == '__main__':
    down = int(input())
    across = int(input())

    for i in range(down):
        matrix.append(list(map(int, input().strip().split())))

    check(1, 1)

    if found:
        print('yes')
    else:
        print('no')
